File: AMap-Android-Kit/app/src/main/java/kain/amap_android_kit/MapOptions.java
package kain.amap_android_kit;

import android.graphics.Color;
import android.location.Location;
import android.util.Log;

import com.amap.api.maps.AMap;
import com.amap.api.maps.AMapOptions;
import com.amap.api.maps.CameraUpdateFactory;
import com.amap.api.maps.UiSettings;
import com.amap.api.maps.model.MyLocationStyle;

import java.util.ArrayList;
import java.util.List;
import java.util.Observable;

public class MapOptions implements AMap.OnMyLocationChangeListener {
    //volatile, 声明变量值的一致性；static,声明变量的唯一性。 volatile同步机制:内存同步
    private static volatile MapOptions instance = null;
    private AMap aMap = null;
    private MyLocationStyle myLocationStyle = null;
    private UiSettings uiSet = null;
    private Location myLocation = null;
    protected List<OnMapOptionsLocationChangeListener> OptionsLocationChangeListeners = null;

    private MapOptions(AMap aMap){
        OptionsLocationChangeListeners  = new ArrayList<OnMapOptionsLocationChangeListener>();
        this.aMap = aMap;
        this.uiSet = aMap.getUiSettings();
        /*aMap.setOnMyLocationChangeListener(new AMap.OnMyLocationChangeListener() {
            @Override
            public void onMyLocationChange(Location location) {
                myLocation = location;
                Log.wtf("Msg:",myLocation.getLatitude() + " " + myLocation.getLongitude());
            }
        });*/
        aMap.setOnMyLocationChangeListener(this);
//        Observable
        //this.showPositionDot();
    }


    public void setMyLocation(Location myLocation) {
        this.myLocation = myLocation;
        this.notifyObservers(this.myLocation);
    }

    /**
     * @param aMap 地图对象
     * @return MapOptions对象
     */
    public static MapOptions getInstance(AMap aMap) {
        // if already inited, no need to get lock everytime
        if (instance == null) {
            synchronized (MapOptions.class) {
                if (instance == null) {
                    instance = new MapOptions(aMap);
                }
            }
        }

        return instance;
    }

    /**
     * 开启默认蓝点
     */
    public void showPositionDot(){
        myLocationStyle = new MyLocationStyle();//初始化定位蓝点样式类myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_LOCATION_ROTATE);//连续定位、且将视角移动到地图中心点，定位点依照设备方向旋转，并且会跟随设备移动。（1秒1次定位）如果不设置myLocationType，默认也会执行此种模式。
        //myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_MAP_ROTATE);
        myLocationStyle.interval(2000); //设置连续定位模式下的定位间隔，只在连续定位模式下生效，单次定位模式下不会生效。单位为毫秒。
        //        精度圈颜色自定义方法如下
        myLocationStyle.strokeColor(Color.BLACK); //设置定位蓝点精度圆圈的边框颜色的方法。
        myLocationStyle.radiusFillColor(Color.argb(127,82, 249, 82));
        aMap.setMyLocationStyle(myLocationStyle);//设置定位蓝点的Style
        //aMap.getUiSettings().setMyLocationButtonEnabled(true);设置默认定位按钮是否显示，非必需设置。
        aMap.setMyLocationEnabled(true);// 设置为true表示启动显示定位蓝点，false表示隐藏定位蓝点并不进行定位，默认是false。
    }

    /**
     * 完全自定义蓝点
     * @param myLocationStyle 自定义蓝点
     */
    public void showPositionDot(MyLocationStyle myLocationStyle) {
        aMap.setMyLocationStyle(myLocationStyle);
        aMap.setMyLocationEnabled(true);
    }

    /**
     * 设置蓝点样式开启蓝点
     * @param LOCATION_TYPE  MyLocationStyle.LOCATION_TYPE 蓝点样式
     */
    public void showPositionDot(int LOCATION_TYPE){
        /*
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_SHOW);//只定位一次。
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_LOCATE) ;//定位一次，且将视角移动到地图中心点。
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_FOLLOW) ;//连续定位、且将视角移动到地图中心点，定位蓝点跟随设备移动。（1秒1次定位）
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_MAP_ROTATE);//连续定位、且将视角移动到地图中心点，地图依照设备方向旋转，定位点会跟随设备移动。（1秒1次定位）
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_LOCATION_ROTATE);//连续定位、且将视角移动到地图中心点，定位点依照设备方向旋转，并且会跟随设备移动。（1秒1次定位）默认执行此种模式。
        //以下三种模式从5.1.0版本开始提供
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_LOCATION_ROTATE_NO_CENTER);//连续定位、蓝点不会移动到地图中心点，定位点依照设备方向旋转，并且蓝点会跟随设备移动。
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_FOLLOW_NO_CENTER);//连续定位、蓝点不会移动到地图中心点，并且蓝点会跟随设备移动。
        myLocationStyle.myLocationType(MyLocationStyle.LOCATION_TYPE_MAP_ROTATE_NO_CENTER);//连续定位、蓝点不会移动到地图中心点，地图依照设备方向旋转，并且蓝点会跟随设备移动。
        * */
        myLocationStyle.myLocationType(LOCATION_TYPE);
        myLocationStyle.interval(2000); //设置连续定位模式下的定位间隔，只在连续定位模式下生效，单次定位模式下不会生效。单位为毫秒。
        //        精度圈颜色自定义方法如下
        myLocationStyle.strokeColor(Color.BLACK); //设置定位蓝点精度圆圈的边框颜色的方法。
        myLocationStyle.radiusFillColor(Color.argb(127,82, 249, 82));
        aMap.setMyLocationStyle(myLocationStyle);//设置定位蓝点的Style
        //aMap.getUiSettings().setMyLocationButtonEnabled(true);设置默认定位按钮是否显示，非必需设置。
        aMap.setMyLocationEnabled(true);// 设置为true表示启动显示定位蓝点，false表示隐藏定位蓝点并不进行定位，默认是false。

    }

    /**
     * 是否开启室内地图
     * @param stuas
     */
    public void showsIndoorMap(Boolean stuas){
        aMap.showIndoorMap(stuas);
    }

    /**
     * 是否启动实时路况
     * @param stuas
     */
    public void showTraffic(Boolean stuas){
        aMap.setTrafficEnabled(stuas);
    }

    /**
     * 地图图层设置
     * @param var1
     * AMap.MAP_TYPE_NAVI 导航地图
     * AMap.MAP_TYPE_NIGHT 夜景地图
     * AMap.MAP_TYPE_SATELLITE 卫星图
     */
    public void setMapType(int var1) {
        aMap.setMapType(var1);
    }

    /**
     * 设置高德地图”Logo的位置。
     * position
     * LOGO_POSITION_BOTTOM_LEFT = 0;
     LOGO_POSITION_BOTTOM_CENTER = 1;
     LOGO_POSITION_BOTTOM_RIGHT = 2;
     ZOOM_POSITION_RIGHT_CENTER = 1;
     ZOOM_POSITION_RIGHT_BUTTOM = 2;
     LOGO_MARGIN_LEFT = 0;
     LOGO_MARGIN_RIGHT = 1;
     LOGO_MARGIN_BOTTOM = 2;
     */
    public void setLogoPosition(int position) {
        uiSet.setLogoPosition(position);
    }

    /**
     * 设置默认高德地图”Logo的位置。
     */
    public void setLogoPosition() {
        uiSet.setLogoPosition(AMapOptions.LOGO_POSITION_BOTTOM_LEFT);
    }

    /**
     * 设置是否开启缩放按钮和缩放按钮位置
     * @param stuas
     * @param position
     * LOGO_POSITION_BOTTOM_LEFT = 0;
    LOGO_POSITION_BOTTOM_CENTER = 1;
    LOGO_POSITION_BOTTOM_RIGHT = 2;
    ZOOM_POSITION_RIGHT_CENTER = 1;
    ZOOM_POSITION_RIGHT_BUTTOM = 2;
    LOGO_MARGIN_LEFT = 0;
    LOGO_MARGIN_RIGHT = 1;
    LOGO_MARGIN_BOTTOM = 2;
     */
    public void setZoomControls(Boolean stuas,int position) {
        uiSet.setZoomControlsEnabled(stuas);
        uiSet.setZoomPosition(position);
    }

    /**设置是否开启缩放按钮
     * @param stuas
     */
    public void setZoomControls(Boolean stuas) {
        uiSet.setZoomControlsEnabled(stuas);
        uiSet.setZoomPosition(AMapOptions.LOGO_POSITION_BOTTOM_LEFT);
    }

    /**
     * 是否显示指南针
     */
    public void setCompass(Boolean stuas) {
        uiSet.setCompassEnabled(stuas);
    }

    /**
     * 是否显示比例尺
     */
    public void setScale(Boolean stuas) {
        uiSet.setScaleControlsEnabled(stuas);
    }

    /**
     * 设置地图显示语言
     * 即"zh_cn", AMap.ENGLISH 表示英文，即"en"
     */
    public void setMapLanguage(String languge) {
        aMap.setMapLanguage(languge);
    }



    /**
     * @param stuas 开|关
     * 设置缩放手势
     */
    public void setZoomGestures(Boolean stuas){
        uiSet.setZoomGesturesEnabled(stuas);
    }

    /**
     * @param stuas 开|关
     * 设置滑动手势
     */
    public void setScrollGestures(Boolean stuas){
        uiSet.setScrollGesturesEnabled(stuas);
    }

    /**
     * @param stuas 开|关
     * 设置旋转手势
     */
    public void setRotateGestures(Boolean stuas){
        uiSet.setRotateGesturesEnabled(stuas);
    }

    /**
     * @param stuas 开|关
     * 设置倾斜手势
     */
    public void setTiltGestures(Boolean stuas){
        uiSet.setTiltGesturesEnabled(stuas);
    }

    /**
     * @param stuas 开|关
     * 设置所有手势
     */
    public void setAllGestures(Boolean stuas){
        uiSet.setAllGesturesEnabled(stuas);
    }

    /**
     * 缩放手势是否开启
     */
    public Boolean isZoomGesturesEnabled(){
        return uiSet.isZoomGesturesEnabled();
    }

    /**
     * 滑动手势是否开启
     */
    public Boolean isScrollGesturesEnabled(){
        return uiSet.isScrollGesturesEnabled();
    }

    /**
     * 旋转手势是否开启
     */
    public Boolean isRotateGesturesEnabled(){
        return uiSet.isRotateGesturesEnabled();
    }

    /**
     * 倾斜手势是否开启
     */
    public Boolean isTiltGesturesEnabled(){
        return uiSet.isTiltGesturesEnabled();
    }


    /**
     * 指定屏幕中心点的手势操作 在对地图进行手势操作时（滑动手势除外），可以指定屏幕中心点后执行相应手势。
     * //x、y均为屏幕坐标，屏幕左上角为坐标原点，即(0,0)点。
     * 当x和y同时为-1时 设置坐标不生效
     * @param x
     * @param y
     * @param stuas 是否启用
     */
    public void setGestureScaleByMapCenter(int x, int y, Boolean stuas){
        if(x==y && x != -1){
            aMap.setPointToCenter(x, y);
        }
        uiSet.setGestureScaleByMapCenter(stuas);
        CameraUpdateFactory.zoomTo(17);
    }

    @Override
    public void onMyLocationChange(Location location) {
        if(location != null){
            this.setMyLocation(location);
        }
    }

    /**
     * 设置监听MapOptionsLocation变化时 回调(注册监听者)
     * @param var1 观察者
     */
    public void setOnMapOptionsLocationChangeListener(MapOptions.OnMapOptionsLocationChangeListener var1){
        OptionsLocationChangeListeners.add(var1);
    }

    /**
     * 定义监听接口
     */
    public interface OnMapOptionsLocationChangeListener {
        void onMapOptionsLocationChange(Location location);
    }

    /**
     * 主题内容发生改变时，会即时通知观察者做出反应
     * @param location
     */
    public void notifyObservers(Location location) {
        for (OnMapOptionsLocationChangeListener OptionsLocationChangeListener : OptionsLocationChangeListeners) {
            OptionsLocationChangeListener.onMapOptionsLocationChange(location);
        }
    }
}

File: AMap-Android-Kit-Kotlin/app/src/main/java/kain/amap_android_kit_kotlin/MapOptions.kt
package kain.amap_android_kit_kotlin

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.location.Location
import android.os.Bundle
import android.os.Environment
import android.util.Log
import com.amap.api.maps.*
import io.reactivex.subjects.AsyncSubject
import com.amap.api.maps.MapView
import com.amap.api.maps.AMapOptions
import com.amap.api.maps.CameraUpdateFactory
import com.amap.api.maps.CameraUpdate
import com.amap.api.maps.model.*
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*


class MapOptions(map: AMap) {
    private var map: AMap? = null
    private var locationStyle: MyLocationStyle? = null
    private var uiSet: UiSettings? = null
    var asyncMapLocation: AsyncSubject<Location> = AsyncSubject.create()
    var asyncScreenShot: AsyncSubject<String> = AsyncSubject.create()


    init {
        this.map = map
        this.uiSet = this.map?.getUiSettings()
        this.map?.setOnMyLocationChangeListener {
            if (it != null) {
                asyncMapLocation.onNext(it)
                asyncMapLocation.onComplete()
            } else {
                Log.e("AmapError", "location Error")
            }
        }


        this.asyncMapLocation.subscribe({
            Log.wtf("Msg", it.latitude.toString() + ":" + it.longitude.toString())
        })

        /*this.asyncScreenShot.subscribe({
            Log.wtf("Msg", "地图截图位置为："+it)
        })*/
    }

    /**
     * TODO:显示定位蓝点
     */
    fun showPositionDot() {
        locationStyle = MyLocationStyle()
        // 连续定位时间间隔
        locationStyle?.interval(2000)
        // 是否显示蓝点
        locationStyle?.showMyLocation(true)
        // 自定义定位图标
        // locationStyle?.myLocationIcon(BitmapDescriptorFactory.fromResource(R.drawable.mark_bs))
        // 定位蓝点精度圆圈的边框颜色
        locationStyle?.strokeColor(Color.RED)
        // 定位蓝点精度圆圈的填充颜色
        locationStyle?.radiusFillColor(Color.BLACK)
        // 精度圈边框宽度自定义
        locationStyle?.strokeWidth((2).toFloat())
        map?.setMyLocationStyle(locationStyle!!)
        // 设置默认定位按钮是否显示
        map?.getUiSettings()?.setMyLocationButtonEnabled(true)
        map?.setMyLocationEnabled(true)
    }

    /**
     * TODO:开启室内地图
     */
    fun showsIndoorMap() {
        map?.showIndoorMap(true)
    }

    /**
     * TODO:显示交通状态
     */
    fun showTraffic() {
        map?.setTrafficEnabled(false)
    }

    /**
     * TODO:地图图层设置
     * MAP_TYPE_NAVI 导航地图
     * MAP_TYPE_NIGHT 夜景地图
     * MAP_TYPE_SATELLITE 卫星图
     */
    fun setMapType() {
        map?.setMapType(AMap.MAP_TYPE_NORMAL)
    }

    /**
     * TODO:地图Logo设置
     */
    fun setLogoPosition() {
        uiSet?.setLogoPosition(AMapOptions.LOGO_POSITION_BOTTOM_LEFT)
    }

    /**
     * TODO:缩放控件
     */
    fun setZoomControls() {
        uiSet?.setZoomControlsEnabled(false)
        uiSet?.setZoomPosition(AMapOptions.LOGO_POSITION_BOTTOM_LEFT)
    }

    /**
     * TODO:指南针
     */
    fun setCompass() {
        uiSet?.setCompassEnabled(false)
    }

    /**
     * TODO:比例尺
     */
    fun setScale() {
        uiSet?.setScaleControlsEnabled(false)
    }

    /**
     * 设置地图显示语言
     * 即"zh_cn", AMap.ENGLISH 表示英文，即"en"
     */
    fun setMapLanguage(languge: String) {
        map?.setMapLanguage(languge)
    }



    /**
     * @param stuas 开|关
     * 设置缩放手势
     */
    fun  setZoomGestures(stuas: Boolean){
        uiSet?.setZoomGesturesEnabled(stuas)
    }

    /**
     * @param stuas 开|关
     * 设置滑动手势
     */
    fun setScrollGestures(stuas: Boolean){
        uiSet?.setScrollGesturesEnabled(stuas)
    }

    /**
     * @param stuas 开|关
     * 设置旋转手势
     */
    fun setRotateGestures(stuas: Boolean){
        uiSet?.setRotateGesturesEnabled(stuas)
    }

    /**
     * @param stuas 开|关
     * 设置倾斜手势
     */
    fun setTiltGestures(stuas: Boolean){
        uiSet?.setTiltGesturesEnabled(stuas)
    }

    /**
     * @param stuas 开|关
     * 设置所有手势
     */
    fun setAllGestures(stuas: Boolean){
        uiSet?.setAllGesturesEnabled(stuas)
    }

    /**
     * 缩放手势是否开启
     */
    fun  isZoomGesturesEnabled(): Boolean?{
        return uiSet?.isZoomGesturesEnabled()
    }

    /**
     * 滑动手势是否开启
     */
    fun isScrollGesturesEnabled(): Boolean?{
        return uiSet?.isScrollGesturesEnabled()
    }

    /**
     * 旋转手势是否开启
     */
    fun  isRotateGesturesEnabled(): Boolean?{
        return uiSet?.isRotateGesturesEnabled()
    }

    /**
     * 倾斜手势是否开启
     */
    fun isTiltGesturesEnabled(): Boolean?{
        return uiSet?.isTiltGesturesEnabled()
    }


    /**
     * 指定屏幕中心点的手势操作 在对地图进行手势操作时（滑动手势除外），可以指定屏幕中心点后执行相应手势。
     * //x、y均为屏幕坐标，屏幕左上角为坐标原点，即(0,0)点。
     * 当x和y同时为-1时 设置坐标不生效
     * @param x
     * @param y
     * @param stuas 是否启用
     */

    /**
     * 改变地图的缩放级别
     * 地图的缩放级别一共分为 17 级，从 3 到 19。数字越大，展示的图面信息越精细。
     */
    fun mapZoomTo(value: Float){
        map?.animateCamera(CameraUpdateFactory.zoomTo(value))
    }

    /**
     * 设置地图显示范围的方法，手机屏幕仅显示设定的地图范围
     * 注意：如果限制了地图显示范围，地图旋转手势将会失效。
     */
    fun setMapStatusLimits(startPointLatitude: Double, startPointLongitude: Double, endPointLatitude: Double, endPointLongitude: Double){
        var startPoint: LatLng?=  LatLng(startPointLatitude, startPointLongitude)
        var endPoint: LatLng? = LatLng(endPointLatitude, endPointLongitude)
        var latLngBounds: LatLngBounds? =  LatLngBounds(startPoint, endPoint);
        map?.setMapStatusLimits(latLngBounds);
    }

    /**
     * 改变地图默认显示区域
     * @param Latitude 纬度 目标位置的屏幕中心点经纬度坐标纬度
     * @param Longitude 经度 目标位置的屏幕中心点经纬度坐标经度
     * @param zoomValue 地图缩放级别
     */
    fun setDefaultMap(Latitude: Double, Longitude: Double, zoomValue: Float) {
        // 定义北京市经纬度坐标（此处以北京坐标为例）
        val centerBJPoint = LatLng(Latitude, Longitude)
        // 设置了一个可视范围的初始化位置
        // CameraPosition 第一个参数： 目标位置的屏幕中心点经纬度坐标。
        // CameraPosition 第二个参数： 目标可视区域的缩放级别
        // CameraPosition 第三个参数： 目标可视区域的倾斜度，以角度为单位。
        // CameraPosition 第四个参数： 可视区域指向的方向，以角度为单位，从正北向顺时针方向计算，从0度到360度
        val animateCamera = CameraUpdateFactory.newCameraPosition(CameraPosition(centerBJPoint, zoomValue, 0f, 0f))
        // 定义一个 MapView 对象，构造方法中传入 mapOptions 参数类
         /*var mapView: MapView = MapView(paramContext, mapOptions)
        this.map = mapView.map
        this.uiSet = map?.uiSettings
        return mapView*/
        map?.animateCamera(animateCamera)

    }

    /**
     * 截图
     */
    fun getMapScreenShot(){
        this.map?.getMapScreenShot(object :AMap.OnMapScreenShotListener{

            override fun onMapScreenShot(bitmap: Bitmap?, arg1: Int) {

                val sdf = SimpleDateFormat("yyyyMMddHHmmss")
                if (null == bitmap) {
                    return
                }
                try {
                    val fileName = Environment.getExternalStorageDirectory().toString()+  "/mapScreenShot_" + sdf.format(Date()) + ".png"
                    val fos = FileOutputStream(fileName)
                    val b = bitmap.compress(Bitmap.CompressFormat.PNG, 100, fos)

                    fos.flush()
                    fos.close()


                    val buffer = StringBuffer()
                    if (b)
                        buffer.append("截屏成功 ")
                    else {
                        buffer.append("截屏失败 ")
                    }
                    if (arg1 != 0)
                        buffer.append("地图渲染完成，截屏无网格")
                    else {
                        buffer.append("地图未渲染完成，截屏有网格")
                    }
                    Log.wtf("Msg", buffer.toString())
                    asyncScreenShot.onNext(fileName)
                    asyncScreenShot.onComplete()

                } catch (e: Exception) {
                    e.printStackTrace()
                   // asyncScreenShot.onError(e)
                }
            }

            override fun onMapScreenShot(p0: Bitmap?) {

            }
        })
    }



    fun setGestureScaleByMapCenter(x: Int, y: Int, stuas: Boolean){
        if(x==y && x != -1){
            map?.setPointToCenter(x, y)
        }
        uiSet?.setGestureScaleByMapCenter(stuas)
    }


    /**
     * 添加默认标记
     * 标题
     */
    fun addMarker(latitude: Double, longitude: Double,title: String){
        var latLng = LatLng(latitude,longitude)
        var markerOption = MarkerOptions()
        markerOption.position(latLng)
        markerOption.title(title)
        map?.addMarker(markerOption)
    }

    /**
     * 添加默认标记
     * 标题和内容
     */
    fun addMarker(latitude: Double, longitude: Double,title: String,des: String){
        var latLng = LatLng(latitude,longitude)
        var markerOption = MarkerOptions()
        markerOption.position(latLng)
        markerOption.title(title)
        markerOption.snippet(des)
        map?.addMarker(markerOption)
    }

    /**
     * 添加标记 标题+内容+自定义图标
     */
    fun addMarker(latitude: Double, longitude: Double,title: String,des: String,icon:Bitmap){
        var latLng = LatLng(latitude,longitude)
        var markerOption = MarkerOptions()
        markerOption.position(latLng)
        markerOption.title(title)
        markerOption.snippet(des)
        markerOption.icon(BitmapDescriptorFactory.fromBitmap(icon))
        map?.addMarker(markerOption)
    }

    /**
     * 添加标记 标题+内容+自定义图标+是否可拖动+是否设置平贴效果
     */
    fun addMarker(latitude: Double, longitude: Double,title: String,des: String,icon:Bitmap, draggable: Boolean, isFlat: Boolean){
        var latLng = LatLng(latitude,longitude)
        var markerOption = MarkerOptions()
        markerOption.position(latLng)
        markerOption.title(title)
        markerOption.snippet(des)
        markerOption.icon(BitmapDescriptorFactory.fromBitmap(icon))
        markerOption.draggable(draggable)
        markerOption.setFlat(isFlat)
        map?.addMarker(markerOption)
    }
}

